### ISSUE TITLE
kv: enforce rangefeed filtering based on MVCCValueHeader flag

### ISSUE THREAD

**Issue by miraradeva** [cockroachdb/cockroach#114313]
**Is your feature request related to a problem? Please describe.**
This is part of enabling [write-time CDC filtering](https://docs.google.com/document/d/1bYCLrgOxb2yjTc5aWpZXl8dSB2w60F57GMUcehDRyzY/edit). In particular, this work completes part 6 of the [Solution Breakdown](https://docs.google.com/document/d/1bYCLrgOxb2yjTc5aWpZXl8dSB2w60F57GMUcehDRyzY/edit#heading=h.ezlbc4xeey2n).

**Describe the solution you'd like**
After we add a flag to the MVCCValueHeader to denote that a given write should be omitted from rangefeeds (#113634), we need to enforce the actual filtering in rangefeeds. Code pointers: `Replica.handleLogicalOpLogRaftMuLocked` and
`CatchUpIterator.CatchUpScan`.

On the SQL side we'll ensure that any transactions that write to internal system tables (e.g. `system.span_configurations`, `system.settings`) are prevented from setting the boolean transaction flag `OmitInRangefeeds` to `true`. As part of this issue, we will also be extra safe and make sure the flag is ignored by internal rangefeeds. 

Jira issue: CRDB-33451

Epic CRDB-13169

**Comment by miraradeva** [comment]
I created a PR (#115788) to do the basic filtering on the rangefeed server side but still haven't addressed the concern about system table rangefeed events being affected by a user transaction with `OmitInRangefeeds = true`. Here's a summary of what we discussed:

1. **On the SQL side**. Since `disable_changefeed_replication` is a session variable, there is no way to prevent it from being set based on what writes come in the future. So there is no real way to not allow the session variable to be set when we're writing to system tables. In the planner, we can reject any writes to system tables if the session variable has previously been turned on, but this can potentially happen for many writes and it doesn't seem like a good user experience. 
2. **On the KV rangefeed server side**. If a write to a system-table has `OmitInRangefeeds = true`, we ignore it and expose the write in rangefeeds anyway. The problem with this approach is that on the KV side we have no good way of identifying if a write is to a system table. 
- We can do something similar to `isWatchedSystemTable` where we exhaustively list all system-table descriptors but this seems error-prone (e.g. updating the list when adding a new system table). In fact, `isWatchedSystemTable` is used only for testing.
- We can use `SpanConfig.RangefeedEnabled` as a proxy for "this is a system range", but `RangefeedEnabled` is already overloaded and may be set even for some non-system tables.
- We can add a separate `SpanConfig` flag to indicate that a range is a system range.
3. **On the KV rangefeed client (and server) side**. The rangefeed client can explicitly opt into filtering based on `OmitInRangefeeds` by setting a flag, let's call it `withFiltering`, in the rangefeed client config. This flag would then be propagated via the `kvpb.RangefeedRequest` to the server, where `OmitInRangefeeds` will be respected only if `withFiltering` is set. There are a couple of ways to do this:
- If `withFiltering` is a property of the rangefeed `Processor`, all registrations for a given processor need to either have filtering on or off. This doesn't work well because a replica has a single processor but may have multiple registrations for consumers with different filtering settings (e.g. for both system and non-system ranges).
- If `withFiltering` is a property of a registration, then a processor needs to handle overlapping events to two registrations, one with filtering and one without filtering. This is doable but will force us to do the `OmitInRangefeeds` filtering at the processor level immediately before dispatching events to different registrations (`PublishToOverlapping`) so we can check each registration's filtering preference. This will require adding `OmitInRangefeeds` to the relevant `MVCCLogicalOp` protos (to propagate it to the processor via `ConsumeLogicalOps`), and changing the registry API (`PublishToOverlapping` needs to be aware of `OmitInRangefeeds`). Assuming this is backportable (we already have other proto changes in this feature), I'll go with this option.
